Skip linker tests requiring symlinks on Windows

kcbanner opened the pull request:
Symlinks aren't available on Windows unless developer mode is enabled, or the user has a certain permission.

Running test-link before this change would result in this:

```
error: AccessDenied
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-debug\lib\zig\std\os\windows.zig:220:32: 0x7ff765cd841f in DeviceIoControl (build.exe.obj)
        .PRIVILEGE_NOT_HELD => return error.AccessDenied,
                               ^
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-debug\lib\zig\std\os\windows.zig:741:9: 0x7ff765cbaa6d in CreateSymbolicLink (build.exe.obj)
    _ = try DeviceIoControl(symlink_handle, FSCTL_SET_REPARSE_POINT, buffer[0..buf_len], null);
        ^
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-debug\lib\zig\std\fs.zig:1997:9: 0x7ff765cba207 in symLinkW (build.exe.obj)
        return os.windows.CreateSymbolicLink(self.fd, sym_link_path_w, target_path_w, flags.is_directory);
        ^
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-debug\lib\zig\std\fs.zig:1957:13: 0x7ff765cba194 in symLink (build.exe.obj)
            return self.symLinkW(target_path_w.span(), sym_link_path_w.span(), flags);
            ^
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-debug\lib\zig\std\fs.zig:96:17: 0x7ff765cbabce in atomicSymLink (build.exe.obj)
        else => return err, // TODO zig should know this set does not include PathAlreadyExists
                ^
C:\cygwin64\home\kcbanner\kit\zig\build-stage3-debug\lib\zig\std\build.zig:3690:9: 0x7ff765cbbf22 in doAtomicSymLinks (build.exe.obj)
```

This adds `requires_symlink` and sets it for the linker tests that use symlinks (just macho currently).

I was trying to find a way to use `skip_non_native` through to the test cases to skip these automatically. However, the way the linker tests are setup in this case have a test step that depends on an `addSharedLibrary` step - and that `addSharedLibrary` step is what crashes, not the test itself.

kubkon commented on test/link.zig:
```diff
@@ -151,10 +154,12 @@ fn addMachOCases(cases: *tests.StandaloneContext) void {
 
     cases.addBuildFile("test/link/macho/weak_library/build.zig", .{
         .build_modes = true,
+        .requires_symlinks = true,
```
I believe all tests require symlinks, just some are disabled by default on Win as the macOS SDK is not being provided.

kubkon commented on test/tests.zig:
```diff
@@ -1037,11 +1037,13 @@ pub const StandaloneContext = struct {
 
         if (features.requires_macos_sdk and !self.enable_macos_sdk) return;
         if (features.requires_stage2 and self.omit_stage2) return;
+        if (features.requires_symlinks and builtin.os.tag == .windows) return;
```
I do not think this is correct. I run these tests locally on both my arm64 and x86 Windows installations with Developer Mode enabled thereby having symlinks capability enabled by default and thus have no problem passing MachO tests on Windows. A better solution to this would be to require the user to declare they have symlinking enabled on Windows via a build flag such as `-Denable-symlinks-windows`. Ideally however we would find a way around the symlinks altogether.

Also please note that testing the MachO linker on Windows (or any foreign host) is of paramount importance to weed out any regressions in cross-linking to MachO from those hosts, i.e., targeting macOS from should just work.

kubkon replied:
I think detecting if symlinks are viable might be tricky = the only thing I can think of is trying to create a dummy symlink and checking for error condition. I believe having a declarative mechanism where the user declares a promise they have symlinks rights like I described above is the way to go. We do the same with Rosetta2 on arm macs, and the way this works it that in `build.zig` in the root of the project, we declare a build option that defaults to false:

```zig
// build.zig
const enable_macos_sdk = b.option(bool, "enable-macos-sdk", "Run tests requiring presence of macOS SDK and frameworks") orelse false;
```

If you trace it all the way to the test harness, this should give you a good idea how this mechanism works.

kubkon commented on lib/std/fs.zig:
```diff
@@ -116,6 +124,34 @@ pub fn atomicSymLink(allocator: Allocator, existing_path: []const u8, new_path:
 }
 
+/// TODO remove the allocator requirement from this API
+pub fn atomicHardLink(allocator: Allocator, existing_path: []const u8, new_path: []const u8) !void {
```
Hmm, this PR is touching too much of Zig's API suface and I am not comfortable in adding any hard link support without prior investigation by me or someone from the core team. Would you mind instead just going back to a simple declarative mechanism like we do with `rosetta`? If not, I'll be happy to take it over and drive it to completion for you. Apologies for chucking out your work on hard links but hopefully my explanation makes sense!

kcbanner replied:
Ok, yea that makes sense - no worries. So to be clear, we just want to skip running the macho linker tests (anything that makes a symlink) if `-Denable-windows-symlinks` is not present?

kcbanner commented:
Removed the hard link approach, updated to flag macho tests and requiring symlinks, and to only run those on windows if `-Denable-windows-symlinks` is specified.

kubkon commented on build.zig:
```diff
@@ -88,6 +88,7 @@ pub fn build(b: *Builder) !void {
         "Whether LLVM has the experimental target arc enabled",
     ) orelse false;
     const enable_macos_sdk = b.option(bool, "enable-macos-sdk", "Run tests requiring presence of macOS SDK and frameworks") orelse false;
+    const enable_symlinks_windows = b.option(bool, "enable-symlinks-windows", "Enable using symlinks on Windows. If this is not set, hardlinks are used.") orelse false;
```
```suggestion
    const enable_symlinks_windows = b.option(bool, "enable-symlinks-windows", "Run tests requiring presence of symlinks on Windows") orelse false;
```

kubkon reviewed (approved):
One minor nit but otherwise I think this is a good temporary solution until we figure out the symlink situation on Windows. Thanks!

kcbanner commented:
Thanks for the edit! Cheers!
